indexing-service/src/main/java/io/druid/indexing/common/task/TaskMonitor.java

```diff
@@ -0,0 +1,502 @@
+      return Futures.immediateFailedFuture(new ISE("TaskMonitore is not running"));
+    }
+    final T task = spec.newSubTask(0);
+    log.info("Submitting a new task[%s] for spec[%s]", task.getId(), spec.getId());
+    indexingServiceClient.runTask(task);
+    incrementNumRunningTasks();
+
+    final SettableFuture<SubTaskCompleteEvent<T>> taskFuture = SettableFuture.create();
+    runningTasks.put(
+        spec.getId(),
+        new MonitorEntry(spec, task, indexingServiceClient.getTaskStatus(task.getId()).getStatus(), taskFuture)
+    );
+
+    return taskFuture;
+  }
+
+  /**
+   * Submit a retry task for a failed spec. This method should be called inside of the
+   * {@link java.util.concurrent.Callable} executed by {@link #taskStatusChecker}.
+   */
+  private void retry(String subTaskSpecId, MonitorEntry monitorEntry, TaskStatusPlus lastFailedTaskStatus)
+  {
+    if (running) {
+      final SubTaskSpec<T> spec = monitorEntry.spec;
+      final T task = spec.newSubTask(monitorEntry.taskHistory.size() + 1);
+      log.info("Submitting a new task[%s] for retrying spec[%s]", task.getId(), spec.getId());
+      indexingServiceClient.runTask(task);
+      incrementNumRunningTasks();
+
+      runningTasks.put(
+          subTaskSpecId,
+          monitorEntry.withNewRunningTask(
+              task,
+              indexingServiceClient.getTaskStatus(task.getId()).getStatus(),
+              lastFailedTaskStatus
+          )
+      );
+    }
+  }
+
+  /**
+   * This method should be called after {@link #stop()} to make sure no additional tasks are submitted.
+   */
+  void killAll()
+  {
+    runningTasks.values().forEach(entry -> {
+      final String taskId = entry.runningTask.getId();
+      log.info("Request to kill subtask[%s]", taskId);
+      indexingServiceClient.killTask(taskId);
+    });
+    runningTasks.clear();
+  }
+
+  void incrementNumRunningTasks()
+  {
+    synchronized (taskCountLock) {
+      numRunningTasks++;
+    }
+  }
+
+  void incrementNumSucceededTasks()
+  {
+    synchronized (taskCountLock) {
+      numRunningTasks--;
+      numSucceededTasks++;
+      log.info("[%d/%d] tasks succeeded", numSucceededTasks, expectedNumSucceededTasks);
+    }
+  }
+
+  void incrementNumFailedTasks()
+  {
+    synchronized (taskCountLock) {
+      numRunningTasks--;
+      numFailedTasks++;
+    }
+  }
+
+  boolean isSucceeded()
+  {
+    synchronized (taskCountLock) {
+      return numSucceededTasks == expectedNumSucceededTasks;
+    }
+  }
+
+  int getNumRunningTasks()
+  {
+    synchronized (taskCountLock) {
+      return numRunningTasks;
+    }
+  }
+
+  Status getStatus()
+  {
+    synchronized (taskCountLock) {
+      return new Status(
+          numRunningTasks,
+          numSucceededTasks,
+          numFailedTasks,
+          numSucceededTasks + numFailedTasks,
+          numRunningTasks + numSucceededTasks + numFailedTasks,
+          expectedNumSucceededTasks
+      );
+    }
+  }
+
+  Set<String> getRunningTaskIds()
```

Comment:
do these need to materialize the streams?

Reply:
The Stream API is useful only when lazy evaluation helps. The return value should always be materialized.